File: java/leetcode_grind/Day653.java
package leetcode_grind;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class Day653 {
    static class Pair<F, S> {
        F key;
        S value;

        Pair(F f, S s) {
            key = f;
            value = s;
        }

        F getKey() {
            return key;
        }

        S getValue() {
            return value;
        }
    }

    // https://leetcode.com/problems/path-with-maximum-probability/description/?envType=daily-question&envId=2024-08-31
    static class Solution1 {
        public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
            Map<Integer, List<Pair<Integer, Double>>> graph = new HashMap<>();
            for (int i = 0; i < edges.length; i++) {
                int u = edges[i][0], v = edges[i][1];
                double pathProb = succProb[i];
                graph.computeIfAbsent(u, k -> new ArrayList<>()).add(new Pair<>(v, pathProb));
                graph.computeIfAbsent(v, k -> new ArrayList<>()).add(new Pair<>(u, pathProb));
            }

            double[] maxProb = new double[n];
            maxProb[start_node] = 1.0;

            PriorityQueue<Pair<Double, Integer>> pq = new PriorityQueue<>(
                    (a, b) -> -Double.compare(a.getKey(), b.getKey()));
            pq.add(new Pair<>(1.0, start_node));

            while (!pq.isEmpty()) {
                Pair<Double, Integer> cur = pq.poll();
                double curProb = cur.getKey();
                int curNode = cur.getValue();

                if (curNode == end_node) {
                    return curProb;
                }

                for (Pair<Integer, Double> nxt : graph.getOrDefault(curNode, new ArrayList<>())) {
                    int nxtNode = nxt.getKey();
                    double pathProb = nxt.getValue();
                    if (curProb * pathProb > maxProb[nxtNode]) {
                        maxProb[nxtNode] = curProb * pathProb;
                        pq.add(new Pair<>(maxProb[nxtNode], nxtNode));
                    }
                }
            }

            return 0d;
        }
    }

}

File: src/leetcode_grind/day_653.rs
// https://leetcode.com/problems/path-with-maximum-probability/description/?envType=daily-question&envId=2024-08-31
pub fn max_probability(
    n: i32,
    edges: Vec<Vec<i32>>,
    succ_prob: Vec<f64>,
    start_node: i32,
    end_node: i32,
) -> f64 {
    let mut path_prob = vec![0.0; n as usize];
    path_prob[start_node as usize] = 1.0;

    for _ in 0..n {
        let mut has_update = false;

        for (i, e) in edges.iter().enumerate() {
            let u = e[0] as usize;
            let v = e[1] as usize;
            let p = succ_prob[i];

            if path_prob[u] < path_prob[v] * p {
                path_prob[u] = path_prob[v] * p;
                has_update = true;
            }
            if path_prob[v] < path_prob[u] * p {
                path_prob[v] = path_prob[u] * p;
                has_update = true;
            }
        }

        if !has_update {
            break;
        }
    }

    path_prob[end_node as usize]
}

pub fn max_probability_ii(
    n: i32,
    edges: Vec<Vec<i32>>,
    succ_prob: Vec<f64>,
    start_node: i32,
    end_node: i32,
) -> f64 {
    use std::collections::HashMap;
    use std::collections::VecDeque;

    let mut adj = HashMap::new();
    for (i, e) in edges.iter().enumerate() {
        let u = e[0];
        let v = e[1];
        let p = succ_prob[i];
        adj.entry(u).or_insert(vec![]).push((v, p));
        adj.entry(v).or_insert(vec![]).push((u, p));
    }

    let mut max_prob = vec![0.0; n as usize];
    max_prob[start_node as usize] = 1.0;

    let mut q = VecDeque::new();
    q.push_back(start_node);

    while !q.is_empty() {
        let u = q.pop_front().unwrap();
        if !adj.contains_key(&u) {
            continue;
        }
        for nei in &adj[&u] {
            let v = nei.0;
            let p = nei.1;
            if max_prob[u as usize] * p > max_prob[v as usize] {
                max_prob[v as usize] = max_prob[u as usize] * p;
                q.push_back(v);
            }
        }
    }

    max_prob[end_node as usize]
}
